include/beast/core/impl/flat_streambuf.ipp

```diff
@@ -0,0 +1,314 @@
+basic_flat_streambuf(basic_flat_streambuf&& other)
+    : detail::empty_base_optimization<
+        allocator_type>(std::move(other.member()))
+{
+    p_ = other.p_;
+    in_ = other.in_;
+    out_ = other.out_;
+    last_ = out_;
+    end_ = other.end_;
+    max_ = other.max_;
+    other.p_ = nullptr;
+    other.in_ = nullptr;
+    other.out_ = nullptr;
+    other.last_ = nullptr;
+    other.end_ = nullptr;
+}
+
+template<class Allocator>
+basic_flat_streambuf<Allocator>::
+basic_flat_streambuf(basic_flat_streambuf&& other,
+        Allocator const& alloc)
+    : detail::empty_base_optimization<
+        allocator_type>(alloc)
+{
+    if(this->member() != other.member())
+    {
+        auto const n = other.size();
+        p_ = alloc_traits::allocate(
+            this->member(), n);
+        in_ = p_;
+        out_ = p_ + n;
+        last_ = out_;
+        end_ = out_;
+        max_ = other.max_;
+        std::memcpy(in_, other.in_, n);
+        return;
+    }
+    p_ = other.p_;
+    in_ = other.in_;
+    out_ = other.out_;
+    last_ = out_;
+    end_ = other.end_;
+    max_ = other.max_;
+    other.p_ = nullptr;
+    other.in_ = nullptr;
+    other.out_ = nullptr;
+    other.last_ = nullptr;
+    other.end_ = nullptr;
+}
+
+template<class Allocator>
+basic_flat_streambuf<Allocator>::
+basic_flat_streambuf(
+        basic_flat_streambuf const& other)
+    : detail::empty_base_optimization<allocator_type>(
+        alloc_traits::select_on_container_copy_construction(
+            other.member()))
+{
+    auto const n = other.size();
+    p_ = alloc_traits::allocate(
+        this->member(), n);
+    in_ = p_;
+    out_ = p_ + n;
+    last_ = out_;
+    end_ = out_;
+    max_ = other.max_;
+    std::memcpy(in_, other.in_, n);
+}
+
+template<class Allocator>
+basic_flat_streambuf<Allocator>::
+basic_flat_streambuf(
+    basic_flat_streambuf const& other,
+        Allocator const& alloc)
+    : detail::empty_base_optimization<
+        allocator_type>(alloc)
+{
+    auto const n = other.size();
+    p_ = alloc_traits::allocate(
+        this->member(), n);
+    in_ = p_;
+    out_ = p_ + n;
+    last_ = out_;
+    end_ = out_;
+    max_ = other.max_;
+    std::memcpy(in_, other.in_, n);
+}
+    
+template<class Allocator>
+template<class OtherAlloc>
+basic_flat_streambuf<Allocator>::
+basic_flat_streambuf(
+    basic_flat_streambuf<OtherAlloc> const& other)
+{
+    auto const n = other.size();
+    p_ = alloc_traits::allocate(
+        this->member(), n);
+    in_ = p_;
+    out_  = p_ + n;
+    last_ = out_;
+    end_ = out_;
+    max_ = other.max_;
+    std::memcpy(in_, other.in_, n);
+}
+
+template<class Allocator>
+template<class OtherAlloc>
+basic_flat_streambuf<Allocator>::
+basic_flat_streambuf(
+    basic_flat_streambuf<OtherAlloc> const& other,
+        Allocator const& alloc)
+    : detail::empty_base_optimization<
+        allocator_type>(alloc)
+{
+    auto const n = other.size();
+    p_ = alloc_traits::allocate(
```

Comment:
We can't use constructor delegation because of how the allocators are specified, but I can factor out the common copy/move code.
willfix